eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -1078,7 +1078,11 @@ object Helpers {
         log.error(s"some htlcs don't have a corresponding timeout transaction: tx=$tx, htlcs=$matchingHtlcs, timeout-txs=$matchingTxs")
       }
       matchingHtlcs.zip(matchingTxs).collectFirst {
+        // HTLC transactions cannot change when anchor outputs is unused, so we directly check the txid
```

Comment:
This function is only used for eclair versions before anchor outputs, where we didn't store the `htlcId` in the DB. It indeed doesn't work in the anchor outputs case, but it's not an issue, since in the anchor outputs case we won't call this function (see `trimmedOrTimedOutHtlcs`).

I had to add this new condition because there is one edge case though that I don't know how to better handle.
When we're not using anchor outputs and broadcasting a `claim-htlc-tx` for the first htlc (`id = 0`), we fall into this case (because we cannot discriminate between "htlcId is 0 because it wasn't set in the DB" and "htlcId is 0 because this is the first htlc"). Because we added RBF to `claim-htlc-txs` even without anchor outputs, we have to handle this new case...

Reply:
It does feel correct though: we're trying to match a mempool transaction (a `Transaction` object) with the set of HTLCs we can claim. We previously could rely on the `txid` which was very simple. But whenever the amounts change and we can't rely on `txid` equality, if it does spend the right input and sends it to the right address, it has to be a different version of the same transaction, right?

Because we're the only one that can sign new versions of `claim-htlc-txs` and we will use the same address as the previous one.

Reply:
> It indeed doesn't work in the anchor outputs case, but it's not an issue, since in the anchor outputs case we won't call this function (see trimmedOrTimedOutHtlcs).

Right!

> we fall into this case (because we cannot discriminate between "htlcId is 0 because it wasn't set in the DB" and "htlcId is 0 because this is the first htlc").

Why haven't we set the placeholder value to `-1` in the codecs to remove this ambiguity?

> It does feel correct though: we're trying to match a mempool transaction (a Transaction object) with the set of HTLCs we can claim. We previously could rely on the txid which was very simple. But whenever the amounts change and we can't rely on txid equality, if it does spend the right input and sends it to the right address, it has to be a different version of the same transaction, right?

I'm not saying it's incorrect, I am saying that the `publicKeyScript` is redundant and that confuses me.

Reply:
> Why haven't we set the placeholder value to -1 in the codecs to remove this ambiguity?

This is a good question, it was done in https://github.com/ACINQ/eclair/pull/1728, but I can't find a discussion on this in the PR. I'll dig further.

Reply:
I know why, because `htlcId` uses a `uint64` codec, so we couldn't use a negative value...

Reply:
I think that ship has sailed unfortunately...because now if we read a tx from the DB that has `htlcId == 0`, we cannot know whether it's unset (and should be updated to `-1`) or if it really is the first htlc. We could only do that in the code change that added that field.

Hopefully we can someday remove this compatibility code entirely (it was introduced 8 months ago, so maybe after one year we can consider that eclair nodes have updated at least once since?).

Reply:
> I think that ship has sailed unfortunately

Dammit, I thought we only used that codec as _readonly_. Well I guess ⛵  indeed.

Reply:
It's a good learning though, I won't repeat that mistake next time!